Eigenvalues

georgii-tishenin opened the pull request:
This is a pull request for the first prototype of eigenvalue extraction. The prototype supports eigenvalue extraction at every time step in EMT and DP domains for topologies containing single-phase R, L, C elements, switches and voltage sources.

georgii-tishenin commented on CMakeLists.txt:
```diff
@@ -59,6 +59,8 @@ if(MSVC)
 	# Silence Visual Studio error C2220
 	add_definitions(-D_SILENCE_STDEXT_ARR_ITERS_DEPRECATION_WARNING)
+	
+	add_compile_options(/bigobj) # TODO: [Georgii] remove this after fixing the issue with MNAEigenvalueExtractor.cpp (error C1128)
```
`add_compile_options(/bigobj)` is already a fix suggested to the [error C1128](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/fatal-error-c1128?view=msvc-170) in the error message. This compile option is only applied to MSVC compiler. However, I see it as a workaround and not a nice solution.

I've started getting the error since [Make MNAEigenvalueExtractor a template class ](https://github.com/sogno-platform/dpsim/pull/262/commits/d5252bf07240b07aef967fc67327eea7fad6dc04) commit.
I've tried refactoring the `MNAEigenvalueExtractor` to simplify and reduce the number of functions in the class:

- [Refactor MNAEigenvalueExtractor](https://github.com/sogno-platform/dpsim/pull/262/commits/0803fd50bb73865698d402035fd33afa1b7f6d6d)
- [Remove specialized implementations of logging functions in MNAEigenvalueExtractor](https://github.com/sogno-platform/dpsim/pull/262/commits/adb2cd3dd2743013bba61a64078da7593084d08f)
- [Move eigenvalue logging to a separate class](https://github.com/sogno-platform/dpsim/pull/262/commits/751bfdf4477c23cef42743a24533639d6ce49e0b)

Further refactoring the `MNAEigenvalueExtractor`, e.g, reducing the number of functions or using two separate classes instead of a template class may help resolving the error without `add_compile_options(/bigobj)`. 

My idea is every time I refactor `MNAEigenvalueExtractor` to remove `add_compile_options(/bigobj)` and check if I still get the [error C1128](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/fatal-error-c1128?view=msvc-170). However, it looks like this TODO is confusing.

I suggest that I replace a TODO with a comment explaining the reason to have `add_compile_options(/bigobj)` :
`# Fixes error C1128 for compiling MNAEigenvalueExtractor with MSVC compiler`

What do you say?

georgii-tishenin commented on dpsim-models/src/MathUtils.cpp:
```diff
@@ -97,6 +97,47 @@ Real Math::realFromVectorElement(const Matrix &mat, Matrix::Index row) {
+  return nonZeroMatrix;
+}
+
+MatrixComp Math::convertToComplex(const Matrix &realEquivalentMatrix) {
```
Thank you, done ✔: [Rename a function in MathUtils ](https://github.com/sogno-platform/dpsim/commit/b0b5cd16c65a11882f220e830b1ff8889c061d8d)

martinmoraga replied:
@georgii-tishenin 

for me is the name of this function confusing. Why RealEquivalentMatrix when in the definition of the function are you mentioning the real and imaginary part of this "real equivalent matrix"?

georgii-tishenin replied:
Jan proposed it, ask him😆. Just kidding.
So, the "real equivalent matrix" or longer version "real equivalent of a complex matrix" is the name I've heard for representing a complex matrix with real numbers. This real equivalent matrix stores real part of a complex matrix in upper left and lower right quadrants, and imaginary part in upper right and lower left quadrants:

$$\begin{bmatrix} Re & -Im \\\\ Im & Re \end{bmatrix}$$

The function `convertRealEquivalentMatrixToComplexMatrix`  converts this "real equivalent matrix" to an actual "complex matrix" with complex numbers. In the code of the function there are variables called `realPart` and `imaginaryPart`. They refer rather to real part and imaginary part of the resultant complex matrix. But the real equivalent matrix also stores them.

Or do you know this "real equivalent matrix" under a different name?

georgii-tishenin replied:
Btw, related to matrices. Currently, I am using Dense matrices in the `MNAEigenvalueExtractor`. Another nice proposal from Jan was to switch to Sparse matrices. I'd like to do it, but rather in a small follow-up PR.

georgii-tishenin commented on dpsim/include/dpsim/Simulation.h:
```diff
@@ -174,6 +176,10 @@ class Simulation : public CPS::AttributeList {
   void setSolverType(Solver::Type solverType = Solver::Type::MNA) {
     mSolverType = solverType;
   }
+  ///
```
Done ✔: [Add doxygen comment for a public function in Simulation. ](https://github.com/sogno-platform/dpsim/pull/262/commits/22d179b9cf19207491ac1ff39a5c01c32cdeca61)

georgii-tishenin commented on dpsim/include/dpsim/Simulation.h:
```diff
@@ -95,6 +95,8 @@ class Simulation : public CPS::AttributeList {
   Bool mInitFromNodesAndTerminals = true;
   /// Enable recomputation of system matrix during simulation
   Bool mSystemMatrixRecomputation = false;
+  ///
```
I struggle to find a more descriptive comment than the name `mIsEigenvalueExtractionEnabled` itself. 
I could add something like  `/// Enables eigenvalue extraction during simulation` though.
Would you still recommend to add a comment given that `mIsEigenvalueExtractionEnabled` is `protected` (not `public`)?

georgii-tishenin replied:
Anyway, done ✔: [Add doxygen comment for a protected data member in Simulation](https://github.com/sogno-platform/dpsim/commit/99e8d81543b34f6de67fdf428f35b9a2c809cd57). Thank you!

georgii-tishenin commented on dpsim-models/include/dpsim-models/EMT/EMT_Ph1_Resistor.h:
```diff
@@ -56,6 +58,14 @@ class Resistor : public MNASimPowerComp<Real>,
+
+private:
+  /// Branch index
+  UInt mBranchIdx;
```
Thank you for the idea!
ChatGPT and me would rather not add private data members to an interface 😃, although C++ does not restrict it.

For example, see [ Learn C++ - Pure virtual functions, abstract base classes, and interface classes](https://www.learncpp.com/cpp-tutorial/pure-virtual-functions-abstract-base-classes-and-interface-classes/) :
> An interface class is a class that has no member variables, and where all of the functions are pure virtual! Interfaces are useful when you want to define the functionality that derived classes must implement, but leave the details of how the derived class implements that functionality entirely up to the derived class.

May be we can find a different solution, e.g., composition over inheritance (as you proposed), or moving mBranchIdx to a base class. But for now I would leave it as it is, if you agree.

georgii-tishenin commented on dpsim-models/src/EMT/EMT_Ph1_Resistor.cpp:
```diff
@@ -113,3 +113,15 @@ void EMT::Ph1::Resistor::mnaCompUpdateVoltage(const Matrix &leftVector) {
+  }
+}
+
+void EMT::Ph1::Resistor::setBranchIdx(UInt i) { mBranchIdx = i; }
```
Thank you for bringing attention to this!

Traditionally, interfaces should not provide function implementations. For example, see [Wikipedia - Interface (object-oriented programming)](https://en.wikipedia.org/wiki/Interface_(object-oriented_programming))
> In [object-oriented programming](https://en.wikipedia.org/wiki/Object-oriented_programming), an interface or protocol type[[a]](https://en.wikipedia.org/wiki/Interface_(object-oriented_programming)#cite_note-1) is a [data type](https://en.wikipedia.org/wiki/Data_type) that acts as an [abstraction](https://en.wikipedia.org/wiki/Abstraction) of a [class](https://en.wikipedia.org/wiki/Class_(computer_science)). It describes a set of [method signatures](https://en.wikipedia.org/wiki/Method_signature), the implementations of which may be provided by multiple [classes](https://en.wikipedia.org/wiki/Class_(computer_programming)) that are otherwise not necessarily related to each other.

However, it seems that the concept of interfaces is evolving and having default implementations of functions in an interface is okay. For example, [in C#](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/types/interfaces)

> Beginning with C# 8.0, an interface may define default implementations for some or all of its members. A class or struct that implements the interface doesn't have to implement members that have default implementations. For more information, see [default interface methods](https://learn.microsoft.com/en-us/dotnet/csharp/advanced-topics/interface-implementation/default-interface-methods-versions).

Personally, I would leave it like that for now and think of an alternative solution, while further working on eigenvalue extraction feature.

georgii-tishenin replied:
Thank you, done ✔: [Rephrase/clean up TODOs related to eigenvalue extraction ](https://github.com/sogno-platform/dpsim/commit/8ac5cc8f283dc3df2ed64df8cc99bb41a0c1dc9d)

georgii-tishenin commented on dpsim-models/src/MathUtils.cpp:
```diff
@@ -97,6 +97,47 @@ Real Math::realFromVectorElement(const Matrix &mat, Matrix::Index row) {
   return mat(row, 0);
 }
 
+// TODO: [Georgii] resize existing matrix instead of creating a new one (performance, memory)
```
Thanks for the hint. Do you mean to reuse it instead of `returnNonZeroElements` function?
`sparseView` removes zero or too small values and returns a `SparseMatrix` (not a `DenseMatrix`). 
The problem I see is that `SparseMatrix`  does not completely get rid of zero values, it stores the non-zero values along with their indeces. For example, if you substract 1.0 from a `SparseMatrix`, it will be substracted also from the zero-valued coefficients of the `SparseMatrix`.

And this is how `returnNonZeroElements` is used in `MNAEigenvalueExtractor:`

```
**mDiscreteEigenvalues = CPS::Math::returnNonZeroElements(discreteEigenvaluesIncludingZeros);
**mEigenvalues = 2.0 / mTimeStep * ((**mDiscreteEigenvalues).array() - 1.0) /
                   ((**mDiscreteEigenvalues).array() + 1.0);
```
In this equation I'd like that 1 is substracted only from non-zero-valued coefficients of `mDiscreteEigenvalues`

martinmoraga reviewed (commented):
@georgii-tishenin  Why do you have the two different classes `EigenvalueCompInterface `and `EigenvalueDynamicCompInterface`. Could these classes be merged?

martinmoraga reviewed (commented):
@georgii-tishenin I think the parameter `mBranchIdx` could be part of the class `EigenvalueCompInterface` or `EigenvalueDynamicCompInterface` because this parameter is a specific parameter of the eigenvalue extraction interface

georgii-tishenin commented:
Hey @martinmoraga,
thank you for looking into the code and your comments! Here are my replies:

1. > @georgii-tishenin Why do you have the two different classes `EigenvalueCompInterface `and `EigenvalueDynamicCompInterface`. Could these classes be merged?

There are dynamic components, like inductor or capacitor, and non-dynamic components, like resistor, switch or voltage source.
 To build the matrices for eigenvalue extraction all components need to stamp a value into the `branchNodeIncidenceMatrix`. For that reason I have introduced `EigenvalueCompInterface`, that needs to be implemented by all components (supported by eigenvalue extraction).
Dynamic components need to stamp a value into `signMatrix` and `discretizationMatrix` additionally. For that reason, I have introduced `EigenvalueDynamicCompInterface`, that inherits from `EigenvalueCompInterface `. This way dynamic components need to implement only `EigenvalueDynamicCompInterface`, and due to inheritance they are forced to implement the functions of   `EigenvalueCompInterface` as well. Hope I could explain the idea alright.

2. > @georgii-tishenin I think the parameter `mBranchIdx` could be part of the class `EigenvalueCompInterface` or `EigenvalueDynamicCompInterface` because this parameter is a specific parameter of the eigenvalue extraction interface

Yes, @dinkelbachjan  actually had a similar comment. This is what I replied:

> Thank you for the idea! ChatGPT and me would rather not add private data members to an interface 😃, although C++ does not restrict it.
> 
> For example, see [ Learn C++ - Pure virtual functions, abstract base classes, and interface classes](https://www.learncpp.com/cpp-tutorial/pure-virtual-functions-abstract-base-classes-and-interface-classes/) :
> 
> > An interface class is a class that has no member variables, and where all of the functions are pure virtual! Interfaces are useful when you want to define the functionality that derived classes must implement, but leave the details of how the derived class implements that functionality entirely up to the derived class.
> 
> May be we can find a different solution, e.g., composition over inheritance (as you proposed), or moving mBranchIdx to a base class. But for now I would leave it as it is, if you agree.

So, if you guys insist, I can do it. But personally I would prefer keeping the interface without private members.

martinmoraga commented:
> inheritance they are forced to implement the functions of `EigenvalueCompInterface` as well. Hope I could explain the idea alright.

@georgii-tishenin 

yes, I know what you mean but C++ doesn't have interfaces "enforced by the language", so I my opinion you can define these classes like normal class definition without any special rules.

Another option is to move this variable (and his setter) to `dpsim/dpsim-models/include/dpsim-models/SimPowerComp.h`

martinmoraga commented on dpsim-models/src/MathUtils.cpp:
@georgii-tishenin 

we have a macro defining the "tolerance" to compare doubles.

https://github.com/sogno-platform/dpsim/blob/0f4d17c6dae7efa5b6c27f6c1f1aca767a50aa20/dpsim/include/dpsim/Definitions.h#L14

georgii-tishenin replied:
nice to know, will use it👍

martinmoraga commented:
@georgii-tishenin 

you could add to the docu which components are supported by the eigenvalue extraction "solver"

georgii-tishenin commented:
> @georgii-tishenin
> 
> you could add to the docu which components are supported by the eigenvalue extraction "solver"

yes, that makes sense👍. Will add that.

georgii-tishenin commented:
> > inheritance they are forced to implement the functions of `EigenvalueCompInterface` as well. Hope I could explain the idea alright.
> 
> @georgii-tishenin
> 
> yes, I know what you mean but C++ doesn't have interfaces "enforced by the language", so I my opinion you can define these classes like normal class definition without any special rules.
> 
> Another option is to move this variable (and his setter) to `dpsim/dpsim-models/include/dpsim-models/SimPowerComp.h`

I personally like the 2nd option you propose better: to move the variable `mBranchIdx` to the base class. Shall I do it?
The only downside I see is that `mBranchIdx` is not used by all of the classes derived from `SimPowerComp`.

georgii-tishenin commented:
> @georgii-tishenin
> 
> Regarding the classes `EigenvalueCompInterface ` and `EigenvalueDynamicCompInterface`. You could declare the functions declared in `EigenvalueDynamicCompInterface` in the class `EigenvalueCompInterface ` as empty virtual functions and not as pure virtual functions. Then you can overwrite these functions in the dynamic components, but the "static components" will call the empty version of these function. E.g.:
> 
> ```
>   virtual void stampSignMatrix(MatrixVar<VarType> &signMatrix,
>                                Complex coeffDP) { };
> ```
> 
> It is only an idea to avoid having different interfaces for one "solver"... what do you think about that?

Thanks for the idea! I think these two options are equivalent, I sort of  like the option with `EigenvalueCompInterface ` and `EigenvalueDynamicCompInterface` better, because it also clearly groups the components, implementing the interfaces. You do not get the grouping with just one interface. But is there any special reason to avoid having two different interfaces?

MarvinTollnitschRWTH commented on dpsim-models/include/dpsim-models/Solver/EigenvalueCompInterface.h:
```diff
@@ -0,0 +1,17 @@
+class EigenvalueCompInterface {
+public:
+  using Ptr = std::shared_ptr<EigenvalueCompInterface>;
+  using List = std::vector<Ptr>;
```
This is done in the other component interfaces as well, but generally I think it is better to be explicit about data types. It might not be clear that EigenvalueCompInterface::List is a vector of shared pointers of exactly the class the type was defined in.

It might be hard to understand since EigenvalueCompInterface::List could be any "list"-type containing anything within the interface class. However, once you do know what it means and use it often it may improve readability. I still vote against it though personally.

georgii-tishenin replied:
thanks for your feedback! I agree with your arguments. In this particular case even the actual type would not read too bad `std::vector<std::shared_ptr<EigenvalueCompInterface>>`. However, if there were no aliases at all, codebase in general would become less readable. So, for now let's just keep consistent with the other component interfaces.

m-mirz commented:
@georgii-tishenin I also prefer interface classes without member variables. One issue that we have in dpsim is a deep inheritance hierarchy, each hierarchy level adding new member variables. 
Due to this we often had discussions on how to avoid yet another diamond shape inheritance which causes confusing members.
We should favor interfaces and composition over inheritance.

georgii-tishenin commented:
@m-mirz, thanks for your feedback! I will keep interfaces without member variables.
I am now refactoring after comments from @martinmoraga and @dinkelbachjan. And will soon convert this draft PR back to standard PR.

georgii-tishenin commented:
Closing this PR as I’ve decided to redesign the class relationships to better align with the overall architecture. A revised implementation will follow in a future PR.
